### ISSUE TITLE
Multiple instances of the same Dashboard

### ISSUE THREAD

**Issue by hhaensel** [GenieFramework/Stipple.jl#10]
For various purposes it sounds like a tempting idea to launch several instances of the same Dashboard. However, currently all ws connections use `Genie.config.webchannels_default_route` as channel. Different instances should have different channels.

I propose to allow for an additional parameter `channel` that can be passed to `Stipple.init()`, which assures that the model as well as the js backend communicate via this channel.
I currently use it for a multi-user scenario, with a load balancer sitting in front that manages security and sets a unique id that cannot be modified by the client.
I will submit a series of PRs for Genie, Stipple and StippleUI to show how this may look like.

**Comment by essenciary** [comment]
@hhaensel Thank you, I merged them, that's awesome stuff! I'm not sure I fully understand the use case but I still find this a valuable addition. 

What do you mean by `launch several instances of the same Dashboard`? 

If you're thinking about multiple users using the same app but controlling their own state (eg the model) we don't need different channels. We need different instances of `Model`. Instead of having `const model = Model()` we'll have 
```julia 
const states = Dict{String,ReactiveModel}() 
```
where the key is the session id. 

Is that what you had in mind?

**Comment by hhaensel** [comment]
That's exactly what I'm doing. But during `states[sessionid] = Stipple.init(...)` the webchannel is setup to be the default channel ( normally "__").
Any change in one of the connected clients will trigger a broadcast to all connected clients, irrespective of the model, they are connected to. Therefore, I also personalise the ws-channel and restrict broadcasting only to the correct user in `setup(model, channel)`.

**Comment by essenciary** [comment]
Ah yes, I forgot that we're broadcasting. My initial thought was to use `message` (https://github.com/GenieFramework/Genie.jl/blob/f8d729722ed2a8a6d91cb365b748398e7d08641a/src/WebChannels.jl#L266) but I was short with time for JuliaCon and went for broadcasting instead. Switching to `message` would require some changes to keep a reference to the web socket client. 

Can you share some code when you get it working please?

**Comment by hhaensel** [comment]
Just a comment on the advantage of this setup:
Implementing user-specific channels keeps the state of an app stable even under usage of different browsers or computers. You can also do co-working if two users sign on with a certain key.

**Comment by hhaensel** [comment]
**A short example (without testing...)**

```
# as example a browser-specific distinction
header2user(header) = header["User-Agent"] |> last

route("/") do
    # identify user from the header
    headers = Genie.Requests.getheaders()
    user = header2user(headers)

    # deliver a user-spcific ui
    ui(user)
end
```
```
function ui(user::AbstractString)
    channel = "__/" * hash(user)
    haskey(models, user) || (models[user] = Stipple.init(MultiUserDasboard(), channel = channel))
    model = models[user]
    dashboard(root(model), class="container", [
        heading("Stipple Multi-user (User: $(model.user))"),
        # etc.            
    ], title = "MultiUserDasboard", channel = channel) |> html
end
```

**Comment by hhaensel** [comment]
For testing, you can also set `header2user(header) = "test-user"` and open two browser windows

**Comment by essenciary** [comment]
I think there's multiple things mixed here: 

1 - the authentication mechanism: you think headers, I think cookie/URL but basically it's the same idea. A session id which can be shared (or placed behind a login). It doesn't really matter how the key is passed into the app. 

2 - you use broadcasting to one client - I would go for messaging one client. Different approaches but I can't see any pros/cons. I was thinking about the time complexity of the operations but in both cases it would be a lookup in a Dict by session id so it's O(1). If we're really picky, `message` involves less computations as it writes directly to the socket while broadcast needs to pick the designated ws client before writing out. That shouldn't make a difference though.   

3 - your workflow is the one indicated above, where you build your logic around the channel string. My workflow would be built around the model instance, something like: 
```julia
channel("/__/:session_id") do 
  ui(models[@params(:session_id)]
end
```

There's many ways of solving this problem but sharing the session works with either. 

---

PS - in a multi user app we also need to attach observables to each instance of model. I wonder how that scales. I would be much better if we could attach the observables to the type `Model` itself, instead of having to attach it to every instance.

**Comment by hhaensel** [comment]
Thanks for merging so quickly!
Let me comment on your points

1 - totally agree, but for my purpose that was the easiest and quickest way, as we have to use a loadbalancer anyway

2/3 - maybe, the logic of your setup is cleaner. I'll think about implementing it that way

And I found that the setup accidentally worked on my private machine but not on when I pushed it to a docker. I had a small error in determining the clients and now I see a ping-pong between JShandlers and my backend handlers.

I will submit a PR with a global variable to switch multi-user-mode on and off, which is off by default until I found a patch. It will probably involve modification of the js handlers.

**Comment by essenciary** [comment]
Thanks, merged! I'll see if I can come up with anything on my side.

**Comment by hhaensel** [comment]
I read a bit about vue. It seems that we would need to either move the watchers into the methods section or to check for even.originalEvent...

**Comment by hhaensel** [comment]
I just submitted PR #14

**Comment by essenciary** [comment]
@hhaensel this seems to work great on the master branch. I'll add a link to a demo video in a few minutes. 

```julia
using Stipple

Base.@kwdef mutable struct Name <: ReactiveModel
  name::R{String} = "Stipple!"
end

hs_models = Dict{String,Name}()

function ui()
  hs_model = get!(hs_models, @params(:session_id), Stipple.init(Name(), channel = String(@params(:session_id))))

  [
    page(
      vm(hs_model), class="container", title="Hello Stipple", partial=true, head_content = Genie.Assets.favicon_support(),
      [
        h1([
          "Hello, "
          span("", @text(:name))
        ])

        p([
          "What is your name? "
          input("", placeholder="Type your name", @bind(:name))
        ])
      ]
    )
  ] |> html
end

route("/") do
  Stipple.Genie.Renderer.redirect("/dashboards/$(rand(10_000:99_999))")
end

route("/dashboards/:session_id", ui)

up(rand((8000:9000)), open_browser=true)
```

**Comment by essenciary** [comment]
Oh no, it doesn't fully work :( I thought it did as some parts do, but it seems the state doesn't get synced with the backend (just the front does).

**Comment by hhaensel** [comment]
Could you submit the Stipple and console logs here?

**Comment by hhaensel** [comment]
Well, I'll have a look tomorrow

**Comment by essenciary** [comment]
I'm also digging into this per the path we discussed. I got to the same point of needing to broadcast the changes to the rest of the clients :-/ 

I will push my changes in a branch.

**Comment by hhaensel** [comment]
If you replace page by dashboard, it already works. I think I found the culprit. I am currently only on my mobile. But no need to search further

**Comment by essenciary** [comment]
I implemented a series of changes in Genie and Stipple to support this, based on the code in your branch. I pushed the changes on Genie master and Stipple master - can you please take a look? I have not included the changes you made in the JS files yet. 

Test code: 

```julia
using Stipple

Base.@kwdef mutable struct Name <: ReactiveModel
  name::R{String} = "Stipple!"
end

hs_models = Dict{String,Name}()

function ui()
  channel = String(@params(:session_id))
  hs_model = if haskey(hs_models, channel)
    hs_models[channel]
  else
    hs_models[channel] = Stipple.init(Name(), channel = channel)
  end

  on(hs_model.name) do _
    # @show "changed"
  end

  [
    page(
      vm(hs_model), class="container", title="Hello Stipple", partial=true, channel=channel,
      [
        h1([
          "Hello, "
          span("", @text(:name))
        ])

        p([
          "What is your name? "
          input("", placeholder="Type your name", @bind(:name))
        ])
      ]
    )
  ] |> html
end

route("/") do
  Stipple.Genie.Renderer.redirect("/dashboards/$(rand(10_000:99_999))")
end

route("/dashboards/:session_id", ui)

up(rand((8000:9000)), open_browser=true)
```

**Comment by essenciary** [comment]
The only issue is that at some point the web sockets connection closes :((((( (at least of FF, I haven't checked other browsers yet). 

My tests so far show: 
1 - one browser window - updates work 2-way perfectly
2 - two browser windows - updates work from browser to julia and the data is synced between browser windows
=> however, attempting to push an update from Julia fails (web socket connection closes)

**Comment by essenciary** [comment]
I checked the currently published version of Stipple and that pretty much doesn't work with multiple windows as it doesn't broadcast the changes to all the connected clients. However, pushing from Julia updates all the clients. 

So the master versions work better (both single and multi user) but it needs more tweaking. It's not a Firefox issue, I tested on Edge and it's the same.

**Comment by essenciary** [comment]
I can't believe it but it actually works
https://www.dropbox.com/s/q30z6ogd791m8o3/stipple_multi_user.mov?dl=0

Debouncing needs some tweaking as the updates are pushed slower than typed.

**Comment by essenciary** [comment]
This is the test code with Genie and Stipple master (dev)

```julia
using Stipple

Base.@kwdef mutable struct Name <: ReactiveModel
  name::R{String} = "Stipple!"
end

hs_models = Dict{String,Name}()

function ui()
  channel = String(@params(:session_id))
  hs_model = if haskey(hs_models, channel)
    hs_models[channel]
  else
    hs_models[channel] = Stipple.init(Name(), channel = channel)
  end

  on(hs_models[channel].name) do _
    @show "changed"
  end

  [
    page(
      vm(hs_model), class="container", title="Hello Stipple", partial=true, channel=channel,
      [
        h1([
          "Hello, "
          span("", @text(:name))
        ])

        p([
          "What is your name? "
          input("", placeholder="Type your name", @bind(:name))
        ])
      ]
    )
  ] |> html
end

route("/") do
  Stipple.Genie.Renderer.redirect("/dashboards/$(rand(10_000:99_999))")
end

route("/dashboards/:session_id", ui)

up(rand((8000:9000)), open_browser=true)
```

**Comment by essenciary** [comment]
If it works for you too, we only need to add the production vue mixin, I think.

**Comment by hhaensel** [comment]
You also found out that somehow the channel got lost in `page()` ...

Besides the mixin, I think I need to port the following points:

- broadcasting happens after `update!(model, field, new, old)`. This leads to strange behaviour if buttons are pressed and two clients are connected: The button sets `model.button` to `true` and the correspoding routine `on(callback, model.button)`,  e.g.
```
on(model.button) do
  sleep(3)
  model.button[] = false
end
```
Only after this routine, `push!()` is called, which then sets the value of the other clients to `true` although `model.button` is already false again.
There are three ways out: a) push the actual value, b) call `update!()` with `@async`, c) inverse the order of push`()` and `update!`. I would prefer c). If someone wants async execution he/she can put the aync in his/her routine. What's your opinion?

- you still stick to `newval = Base.parse(valtype, payload["newval"])`. But `payload["newval"]` is already parsed by JSON and may be an `Int64`, although it needs to be parsed as Float64. In that case Base.parse`()`will fails and the type of `newval` will be `Int64`, while `oldval` might be `Float64`. In that case, `update!()` fails, because it requires `oldval` and `newval` to be of the same type.
I still propose to call `convert()`, which is even a lot faster. And to use `Base.parse(valtype, string(payload["newval"]))` as a fallback.

I hope to put an example app here tomorrow.

**Comment by essenciary** [comment]
@hhaensel Yes, I'm not a fan of how we need to pass `channel` in so many places (with `init`, with `page`). It's easy to forget so error prone. Once we release this feature it's worth refactoring the API a bit. The easiest thing would be to check for a `channel` property on the model instance - though this approach is not idiomatic Julia. 

Re first point I didn't realise it was an issue. I like the `@async` idea and tried it but ended up with connection failures in the HTTP server. Might be worth digging into, to confirm - but I suspect it's because `@async` creates a new task and `@params` is no longer accessible. If `c)` solves it, that sounds best. 

Regarding the second point, again, I didn't stumble into this one but happy to go with your solution. 

Would you like to add these into master or should I do it?

**Comment by hhaensel** [comment]
I can do it tomorrow

**Comment by hhaensel** [comment]
Pushed my commits.
My test case is
```# Stipple x-y Line Plot
using Stipple
using StippleCharts, StippleCharts.Charts
using StippleUI, StippleUI.Range, StippleUI.Heading, StippleUI.Dashboard, StippleUI.FormInput
using Genie, Genie.Router, Genie.Renderer.Html

import Genie.Renderer.Json.JSONParser.JSONText

OptDict = Dict{Symbol, Any}
opts(;kwargs...) = OptDict(kwargs...)

const plot_opts_dict = OptDict(
    :chart => OptDict(:type => :line),
    :xaxis => OptDict(:type => :numeric),
    :yaxis => OptDict(:min => -5, :max => 5, :tickAmount => 10,
                      :labels => OptDict(:formatter => JSONText("function(val, index) { return val.toFixed(1); }"))
              )
)

xx = Base.range(0, 4π, length=200) |> collect

Base.@kwdef mutable struct HHDashboard <: ReactiveModel
    name::R{String} = "World"
    a::R{Float64} = 1.0
    b::R{Float64} = 0.0
    c::R{Float64} = 0.0
    plot_data::R{Vector{PlotSeries}} = [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    plot_options_dict::OptDict = plot_opts_dict
end

Stipple.register_components(HHDashboard, StippleCharts.COMPONENTS)
# model = Stipple.init(HHDashboard())
hs_models = Dict{String, HHDashboard}()


function ui()
    @info @params
    channel = String(@params(:session_id))
    model = if haskey(hs_models, channel)
        hs_models[channel]
    else
        hs_models[channel] = Stipple.init(HHDashboard(), channel = channel)
    end
    # update plot_data when a, b or c are changed
    Base.map!(model.plot_data, model.a, model.b, model.c) do a, b, c
        @info "amplitude: $a, phase: $b, offset: $c"
        return [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    end

    dashboard(root(model), class="container", [
        heading("Stipple x-y Line Plot"),
        row(
            cell(class="st-module", h2(["Hello ", span("", @text(:name)), "!"]))
        ),
        row(
            cell(class="st-module", row([
                cell(class="st-br", style="padding-right: 20px;", [
                    h5("What is your name?"),
                    textfield("", :name, placeholder="type your name", label="Name", outlined="", filled="")
                ]),
                cell(class="st-br", style="padding-left: 20px;", [
                    h5("Sine oder Cosine?"),
                    row([
                        cell(size=4, h6("Amplitude")),
                        cell(slider( 0:0.01:5, @data(:a); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Phase")),
                        cell(slider( 0:0.01:5, @data(:b); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Offset")),
                        cell(slider( -3:0.01:3, @data(:c); markers=true, label=true))
                    ])
                ])
            ]))
        ),
        row(
            cell(class="st-module", plot(:plot_data; options=:plot_options_dict))
        ),
        # make a nice bottom section
        row("&nbsp;")
    ], title = "Stipple x-y ApexChart", channel = channel) |> html
end

route("/") do
  Stipple.Genie.Renderer.redirect("/dashboards/$(rand(10_000:99_999))")
end

route("/dashboards/:session_id", ui)

up(open_browser = true)

```

**Comment by hhaensel** [comment]
@essenciary I needed to add Stipple 0.7 to compat of StippleCharts. Could make a new version and set the compat?

**Comment by essenciary** [comment]
@hhaensel I've already made some updates but I haven't pushed yet. Here is what I added: 
![image](https://user-images.githubusercontent.com/11292378/97586901-15e8dc00-19fb-11eb-970e-196c8551869b.png)

If you have the same, yes please. I've sent you an invite for the StippleCharts repo.

**Comment by hhaensel** [comment]
Sounds good, please push.

Have you tested from your side? And have you tested other apps?

**Comment by essenciary** [comment]
Will test a bit later and if all good register Stipple @ v0.7. Once that is released I'll push and register StippleUI and StippleCharts 😊

P.S. - managing and versioning so many dependent packages is pretty messy. Thinking of a tool (maybe using GitHub actions) to automate these.

**Comment by hhaensel** [comment]
If you think about refactoring, I recommend to have introduce private properties, that are not shared with the browser. One could, e.g., decide that fields starting with a _ are not broadcasted nor watched

**Comment by essenciary** [comment]
@hhaensel That's a great point, we should allow private properties. Related to this, I also thought about read-only properties (that can only be updated one way, eg from backend to front, to implement security). Maybe we can define some wrapper types `Private{...}`, `ReadOnly{...}`, etc?

**Comment by hhaensel** [comment]
That's exactly what I had in mind.

I just pushed a slight modification to Stipple and requested a [PR for Genie](https://github.com/GenieFramework/Genie.jl/pull/297) in order to make Stipple and Genie consistent

**Comment by hhaensel** [comment]
I think, we could put the mixin in a general place. Would be stipplecore.js a good location for that?

**Comment by hhaensel** [comment]
Multiple users now works like a charm and the sources are nicely organised 😄 
![grafik](https://user-images.githubusercontent.com/31985040/97641121-7bac8680-1a42-11eb-888b-2f95073a67d9.png)

**Comment by hhaensel** [comment]
This is my latest testcase:
```using Stipple
using StippleCharts, StippleCharts.Charts
using StippleUI, StippleUI.Range, StippleUI.Heading, StippleUI.Dashboard, StippleUI.FormInput
using Genie, Genie.Router, Genie.Renderer.Html

import Genie.Renderer.Json.JSONParser.JSONText

OptDict = Dict{Symbol, Any}
opts(;kwargs...) = OptDict(kwargs...)

const plot_opts_dict = OptDict(
    :chart => OptDict(:type => :line),
    :xaxis => OptDict(:type => :numeric),
    :yaxis => OptDict(:min => -5, :max => 5, :tickAmount => 10,
                      :labels => OptDict(:formatter => JSONText("function(val, index) { return val.toFixed(1); }"))
              )
)

xx = Base.range(0, 4π, length=200) |> collect

Base.@kwdef mutable struct HHDashboard <: ReactiveModel
    name::R{String} = "World"
    a::R{Float64} = 1.0
    b::R{Float64} = 0.0
    c::R{Float64} = 0.0
    plot_data::R{Vector{PlotSeries}} = [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    plot_options_dict::OptDict = plot_opts_dict
end

Stipple.register_components(HHDashboard, StippleCharts.COMPONENTS)
# model = Stipple.init(HHDashboard())
models = Dict{String, HHDashboard}()


function ui(user)
    channel = string(hash(user))

    model = if haskey(models, channel)
        models[channel]
    else
        models[channel] = Stipple.init(HHDashboard(), channel = channel)
    end
    # update plot_data when a, b or c are changed
    Base.map!(model.plot_data, model.a, model.b, model.c) do a, b, c
        @info "amplitude: $a, phase: $b, offset: $c"
        return [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    end

    dashboard(root(model), class="container", [
        heading("Stipple x-y Line Plot"),
        row(
            cell(class="st-module", [h2(["Hello ", span("", @text(:name)), "!"]), p("I am $user")])
        ),
        row(
            cell(class="st-module", row([
                cell(class="st-br", style="padding-right: 20px;", [
                    h5("What is your name?"),
                    textfield("", :name, placeholder="type your name", label="Name", outlined="", filled="")
                ]),
                cell(class="st-br", style="padding-left: 20px;", [
                    h5("Sine oder Cosine?"),
                    row([
                        cell(size=4, h6("Amplitude")),
                        cell(slider( 0:0.01:5, @data(:a); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Phase")),
                        cell(slider( 0:0.01:5, @data(:b); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Offset")),
                        cell(slider( -3:0.01:3, @data(:c); markers=true, label=true))
                    ])
                ])
            ]))
        ),
        row(
            cell(class="st-module", plot(:plot_data; options=:plot_options_dict))
        ),
        # make a nice bottom section
        row("&nbsp;")
    ], title = "Stipple x-y ApexChart", channel = channel) |> html
end

route("/") do
    # identify user from the header
  headers = Genie.Requests.getheaders()
  user = headers["User-Agent"] |> split |> last
  # deliver a user-spcific ui
  ui(user)
end

up(open_browser = true)

```

**Comment by essenciary** [comment]
All right - pulling all the masters and putting my tester hat! Will run through all the demos. 🚀

**Comment by essenciary** [comment]
@hhaensel I've been doing some testing and it looks like the switch from `parse` to `convert` is breaking some features. I'm debugging it atm and we'll merge once I figure it out.

**Comment by hhaensel** [comment]
I very recently realised that you had define your own parse routine, and I confused Base.parse with Meta.parse.
Most probably things will work fine if you eliminate `string()` from the fallback.
No time for testing right now. Will look into it this evening, if you haven't solved it by then 😉

**Comment by essenciary** [comment]
I'm working on it - I think I'll just reverse to `parse` given that it's used as the extension mechanism in StippleUI and StippleCharts. Unless there's a clear benefit with using convert? Sadly for now the package won't precompile and for the life of me I can't tell why :-O

**Comment by hhaensel** [comment]
Could you paste a minimal example of where it goes wrong?

**Comment by hhaensel** [comment]
Anyhow, inverting the order should be ok for me, because I encountered an error during Base.parse.

**Comment by essenciary** [comment]
@essenciary Fixed the issue - everything is merged except for the `convert` instead of `parse`. That would require updating all the `parse` methods in UI and Charts. We can do it later if at all. 

I released a new Genie 1.8 which includes the support for the new Stipple features. Stipple release coming up. Then UI and Charts :)

**Comment by hhaensel** [comment]
Great you now released all the versions 👍 

I still wonder, whether it would a good idea to have the `convert()` in the catch part. So something like

```newval = try
      Base.parse(valtype, payload["newval"])
    catch _
      try
        convert(valtype, payload["newval"])
      catch ex
        @error ex
        payload["newval"]
      end
    end
```
Tomorrow I will find some time for testing.

**Comment by hhaensel** [comment]
I think the root cause for the problem is that Integers are not part of the Reals, so `Int64 <: Float64` is `false`
But I found a cure for my problem which reads:
```
Base.parse(::Type{Float64}, v::T) where {T<:Real} = convert(Float64, v)
```

**Comment by hhaensel** [comment]
Finally, I paste my latest test file here. If you leave out the definition of `Base.parse()` you receive an error as soon as the old value or the new value of a slider is an integer.
With this, I think, we can close this issue 😄 
```
using Stipple
using StippleCharts, StippleCharts.Charts
using StippleUI, StippleUI.Range, StippleUI.Heading, StippleUI.Dashboard, StippleUI.FormInput
using Genie, Genie.Router, Genie.Renderer.Html

import Genie.Renderer.Json.JSONParser.JSONText

Base.parse(::Type{Float64}, v::T) where {T<:Real} = convert(Float64, v)

OptDict = Dict{Symbol, Any}
opts(;kwargs...) = OptDict(kwargs...)

const plot_opts_dict = OptDict(
    :chart => OptDict(:type => :line),
    :xaxis => OptDict(:type => :numeric),
    :yaxis => OptDict(:min => -5, :max => 5, :tickAmount => 10,
                      :labels => OptDict(:formatter => JSONText("function(val, index) { return val.toFixed(1); }"))
              )
)

xx = Base.range(0, 4π, length=200) |> collect

Base.@kwdef mutable struct HHDashboard <: ReactiveModel
    name::R{String} = "World"
    a::R{Float64} = 1.0
    b::R{Float64} = 0.0
    c::R{Float64} = 0.0
    plot_data::R{Vector{PlotSeries}} = [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    plot_options_dict::OptDict = plot_opts_dict
end

Stipple.register_components(HHDashboard, StippleCharts.COMPONENTS)
# model = Stipple.init(HHDashboard())
models = Dict{String, HHDashboard}()


function ui(user)
    channel = string(hash(user))

    model = if haskey(models, channel)
        models[channel]
    else
        model = models[channel] = Stipple.init(HHDashboard(), channel = channel)

        # update plot_data when a, b or c are changed
        Base.map!(model.plot_data, model.a, model.b, model.c) do a, b, c
            @info "amplitude: $a, phase: $b, offset: $c"
            return [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
        end
        model
    end

    dashboard(root(model), class="container", [
        heading("Stipple x-y Line Plot"),
        row(
            cell(class="st-module", [h2(["Hello ", span("", @text(:name)), "!"]), p("I am $user")])
        ),
        row(
            cell(class="st-module", row([
                cell(class="st-br", style="padding-right: 20px;", [
                    h5("What is your name?"),
                    textfield("", :name, placeholder="type your name", label="Name", outlined="", filled="")
                ]),
                cell(class="st-br", style="padding-left: 20px;", [
                    h5("Sine oder Cosine?"),
                    row([
                        cell(size=4, h6("Amplitude")),
                        cell(slider( 0:0.01:5, @data(:a); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Phase")),
                        cell(slider( 0:0.01:5, @data(:b); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Offset")),
                        cell(slider( -3:0.01:3, @data(:c); markers=true, label=true))
                    ])
                ])
            ]))
        ),
        row(
            cell(class="st-module", plot(:plot_data; options=:plot_options_dict))
        ),
        # make a nice bottom section
        row("&nbsp;")
    ], title = "Stipple x-y ApexChart", channel = channel) |> html
end

route("/") do
  # identify user from the header
  headers = Genie.Requests.getheaders()
  user = headers["User-Agent"] |> split |> last
  # deliver a user-spcific ui
  ui(user)
end

up(open_browser = true)
```

**Comment by essenciary** [comment]
@hhaensel Should we extend with the `parse` method into StippleUI or directly in Stipple? Eg: `Base.parse(::Type{Float64}, v::T) where {T<:Real} = convert(Float64, v)`. 

If we put it in Stipple, I'm not sure it's always desirable. If we put it into packages, multiple packages might declare it, leading to to conflicts (and it would count as type piracy, I guess).

**Comment by hhaensel** [comment]
I was also unsure whether we should put in StippleUI for pretty much the same reasons that you pointed out. Did you test my proposal above?

>     newval = try
>       Base.parse(valtype, payload["newval"])
>     catch _
>       try
>         convert(valtype, payload["newval"])
>       catch ex
>         @error ex
>         payload["newval"]
>       end
>     end

**Comment by essenciary** [comment]
@hhaensel Thank you. The `try/catch` approach works but I'm not a fan. Exceptions are expensive computationally and this will be triggered a lot - and it's really a certainty that it will happen, it's not really an _exception_. So we should address it in a scalable and reliable manner, I think.

**Comment by hhaensel** [comment]
I understand your point. I see three ways
- use my previous workaround and integrate it in Stipple
- if we want to avoid type piracy we can check the float types manually with `typeof`
- we should understand why it fails for your use case and provide a solution

For the last point I'd love to see a minimum example. It's probably linked to a certain type you are using.

**Comment by hhaensel** [comment]
Finally, I paste my latest test file here. If you leave out the definition of `Base.parse()` you receive an error as soon as the old value or the new value of a slider is an integer.
With this, I think, we can close this issue 😄 
```
using Stipple
using StippleCharts, StippleCharts.Charts
using StippleUI, StippleUI.Range, StippleUI.Heading, StippleUI.Dashboard, StippleUI.FormInput
using Genie, Genie.Router, Genie.Renderer.Html

import Genie.Renderer.Json.JSONParser.JSONText

Base.parse(::Type{Float64}, v::T) where {T<:Real} = convert(Float64, v)

OptDict = Dict{Symbol, Any}
opts(;kwargs...) = OptDict(kwargs...)

const plot_opts_dict = OptDict(
    :chart => OptDict(:type => :line),
    :xaxis => OptDict(:type => :numeric),
    :yaxis => OptDict(:min => -5, :max => 5, :tickAmount => 10,
                      :labels => OptDict(:formatter => JSONText("function(val, index) { return val.toFixed(1); }"))
              )
)

xx = Base.range(0, 4π, length=200) |> collect

Base.@kwdef mutable struct HHDashboard <: ReactiveModel
    name::R{String} = "World"
    a::R{Float64} = 1.0
    b::R{Float64} = 0.0
    c::R{Float64} = 0.0
    plot_data::R{Vector{PlotSeries}} = [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
    plot_options_dict::OptDict = plot_opts_dict
end

Stipple.register_components(HHDashboard, StippleCharts.COMPONENTS)
# model = Stipple.init(HHDashboard())
models = Dict{String, HHDashboard}()


function ui(user)
    channel = string(hash(user))

    model = if haskey(models, channel)
        models[channel]
    else
        model = models[channel] = Stipple.init(HHDashboard(), channel = channel)

        # update plot_data when a, b or c are changed
        Base.map!(model.plot_data, model.a, model.b, model.c) do a, b, c
            @info "amplitude: $a, phase: $b, offset: $c"
            return [PlotSeries("Sine", PlotData(zip(xx, a .* sin.(xx .- b) .+ c) |> collect))]
        end
        model
    end

    dashboard(root(model), class="container", [
        heading("Stipple x-y Line Plot"),
        row(
            cell(class="st-module", [h2(["Hello ", span("", @text(:name)), "!"]), p("I am $user")])
        ),
        row(
            cell(class="st-module", row([
                cell(class="st-br", style="padding-right: 20px;", [
                    h5("What is your name?"),
                    textfield("", :name, placeholder="type your name", label="Name", outlined="", filled="")
                ]),
                cell(class="st-br", style="padding-left: 20px;", [
                    h5("Sine oder Cosine?"),
                    row([
                        cell(size=4, h6("Amplitude")),
                        cell(slider( 0:0.01:5, @data(:a); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Phase")),
                        cell(slider( 0:0.01:5, @data(:b); markers=true, label=true))
                    ]),
                    row([
                        cell(size=4, h6("Offset")),
                        cell(slider( -3:0.01:3, @data(:c); markers=true, label=true))
                    ])
                ])
            ]))
        ),
        row(
            cell(class="st-module", plot(:plot_data; options=:plot_options_dict))
        ),
        # make a nice bottom section
        row("&nbsp;")
    ], title = "Stipple x-y ApexChart", channel = channel) |> html
end

route("/") do
  # identify user from the header
  headers = Genie.Requests.getheaders()
  user = headers["User-Agent"] |> split |> last
  # deliver a user-spcific ui
  ui(user)
end

up(open_browser = true)
```

**Comment by essenciary** [comment]
@hhaensel Thanks for the example, very useful for the tests :) 

I've been digging into this and type piracy is a big issues, so that's a no go. And with the computational costs of exceptions being high, I came up with this solution: 
https://github.com/GenieFramework/Stipple.jl/commit/e78440f9f478a625f14f12fdd87ec72abc68c2d2

Julia is very fast at checking types so we can use that to our advantage. So I just added explicit support for Int to Float conversions, without extending `parse` or `convert` (so no type piracy). Seems to work great so I registered v0.7.1 with the patch.

**Comment by essenciary** [comment]
@hhaensel Btw, would you like to contribute this example to the demo repo? https://github.com/GenieFramework/StippleDemos

**Comment by hhaensel** [comment]
I think we should close this long issue and address everything else in new issues.
Happy to see it all working 😃 🚀

**Comment by essenciary** [comment]
@hhaensel Yeeee! Throwing one more thing here in case we want to create a new issue for it. What is your take on Jupyter/Pluto integration?

**Comment by hhaensel** [comment]
I was not aware of Pluto - now I am ...
For me, the use case of Stipple was to deploy an app without having the user to bother with any line of code. This is the contrary of Pluto/Jupyter.

I could well imagine that stipple apps live within a notebook cell, but I am probably not the one to implement this 😉

**Comment by essenciary** [comment]
It could open interesting workflows, like no deployments at all, by coding directly on the server/cloud, cloning+editing+republishing a public notebook, etc. I have it on my todo list but indeed, it's a different approach.

**Comment by hhaensel** [comment]
> 
> 
> @hhaensel That's a great point, we should allow private properties. Related to this, I also thought about read-only properties (that can only be updated one way, eg from backend to front, to implement security). Maybe we can define some wrapper types `Private{...}`, `ReadOnly{...}`, etc?

@essenciary 
I thought about this again. What do you think, if we establish the convention that variables ending with underscore are read-only? That's pretty easy to realise. It's just a
```
endswith(payload["field"], "_") && return nothing
```
in the channel watcher and it would prevent us from defining get and set functions for the wrapper.
Ending with two underscores could be private. This is not so easy to change in the code, though, because you would have to alter all occurrences of the variable...

**Comment by essenciary** [comment]
@hhaensel In my opinion, it's better long term to use wrappers (more robust and more Julian). I can look into it these days, as we're looking towards v1, it's worth putting some time into it.
